### ISSUE TITLE
Invalid plugin configuration after migration from 0.14 to 1.1.2 running Cassandra

### ISSUE THREAD

**Issue by chris-branch** [Kong/kong#4631]
When updating from 0.14 to 1.1.2, there are several database schema/data changes that are incomplete and end up leaving several database tables (`plugins`, `consumers`, `upstreams`, and `targets`) in an invalid state.  Three issues:

1) In the `plugins`, `consumers`, `upstreams`, and `targets` tables, existing data needs the `created_at` column truncated to account for recent changes between milliseconds vs. ~micro~ seconds.
2) In the `plugins` table, a new column named `run_on` is added, but no default value is set for existing plugins, so existing plugins end up with `null` for this column.
3) In the `plugins` table, a new column named `protocols` is added, but no default value is set for existing plugins, so existing plugins end up with `null` for this column.

The symptom is that the upgrade will complete successfully, and kong will run successfully, but subsequent attempts to modify the configuration via HTTP PATCH will fail with a 400 response.  For example, updating the config for the CORS plugin:

Request: `PATCH http://localhost:8001/plugins/b5c81c6f-75f4-4639-92aa-8e6694c66e9a`
```
{
    "config": {
        "origins": [
            "example.com"
        ]
    }
}
```
Response: Status = 400
```
{
    "message": "3 schema violations (created_at: expected an integer; protocols: required field missing; run_on: required field missing)",
    "name": "schema violation",
    "fields": {
        "run_on": "required field missing",
        "created_at": "expected an integer",
        "protocols": "required field missing"
    },
    "code": 2
}
```

You will continue to get these errors when attempting to update configuration until the values in the database columns are corrected.  This only affects plugins that exist prior to the upgrade.  New plugins that are created after the upgrade don't have this issue.  The other interesting thing to note is that a GET request to retrieve plugin data will work successfully and automatically cleanup bad/missing values in the response data returned (e.g., you can't see the null values for `run_on` or `protocols` unless you query the database directly).

There are a couple changes needed.  First, it looks like the migration scripts were updated recently to truncate "created_at" for Postgres, but there wasn't any code added for Cassandra.  Cassandra needs an equivalent fix:

https://github.com/Kong/kong/issues/4183
https://github.com/Kong/kong/pull/4196

Second, the two new columns (`run_on` and `protocols`) need default values since these are required columns according to the plugin schema.  The migration scripts should fill in sane defaults for existing plugins as part of the upgrade process because leaving these set to `null` is not valid per the plugin schema.

**Comment by HrushikeshBodas** [comment]
Id love to work on this! How can I get started?

**Comment by bungle** [comment]
@chris-branch,

Thank you for reporting. I think the issue is partially solved here:
https://github.com/Kong/kong/pull/4635

The `milliseconds` and `seconds` issue needs more work, @HrushikeshBodas thank you for offering your hand, would you like to take look on this time issue? @chris-branch is that issue with `Cassandra` or `PostgreSQL` (perhaps both)?

**Comment by kikito** [comment]
Hello @chris-branch , it seems that #1106 might have fixed the problem you are facing. At the same time the 1.4.x branch does not use the `run_on` field any more. Could you try Kong 1.4.3?

**Comment by chris-branch** [comment]
After the upgrade from 0.14 -> 1.1.2 last year, I ended up running a one-off script to fill in default values in the database where needed, so that was how I got around the issue at the time.  I'm currently running 1.4.2 and have not run into any more issues with inconsistencies with plugin schemas vs. database configuration.  In fact, I just tested with 1.4.2, and I'm able to modify the code for one of my plugins to remove several items from the plugin's schema, then successfully GET and PATCH existing plugin configurations that include the now-deleted schema elements.  In earlier versions, Kong would generate a schema error if the config in the database contains extra fields not listed in the plugin's schema (which made it impossible to ever remove schema elements without deleting/re-creating the plugin configs).  In 1.4.2, it quietly removes the invalid items during GET / PATCH.  So, this is working perfectly in 1.4.2 from my perspective.

My upgrade history over the past year is as follows: 0.12 -> 0.14 -> 1.1.2 -> 1.2.1 -> 1.4.2
